### PR TITLE
feat: add configurable custom mapping to dd exporter

### PR DESCRIPTION
**Pull request by samuelAndalon**
*Description here*
This PR fixes the issue with DD exporter not providing meaningful data in the DD traces.
[There is a known issue](https://docs.rs/opentelemetry-datadog/latest/opentelemetry_datadog/#quirks) where open telemetry is not fully compatible with datadog

causing this
image

to fix, this, open-telemetry-datadog [added custom mapping functions](https://docs.rs/opentelemetry-datadog/0.6.0/opentelemetry_datadog/struct.DatadogPipelineBuilder.html#method.with_resource_mapping) 

after applying mapping for `span_name` and `resource_name` DD spans look like this (tested the change in our router implementation)
image

1. **supergraph** span resource name will be the name of the operation sent to the router, example: `myAwesomeQuery`
2. **subgraph** span resource name will be the name of the subgraph that router will be sending a query from query plan, example: `my-federated-subgraph`
3. **subgraph_request** span resouce name will be the name of the operation that router will send to subgraph taken from the query plan, example: `myAwesomeQuery__my-my-federated-subgraph__0`

All this logic is gated behind a yaml configuration boolean `enable_span_mapping` which if enabled will take the values from the span attributes.

Fixes #issue_number
https://github.com/apollographql/router/issues/2228

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [X] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [X] Unit Tests
    - [X] Integration Tests
    - [X] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

File: apollo-router/src/plugins/telemetry/tracing/datadog.rs [hunk 1]
```diff
@@ -35,9 +61,27 @@ impl TracingConfigurator for Config {
             AgentEndpoint::Default(_) => None,
             AgentEndpoint::Url(s) => Some(s),
         };
+        let enable_span_mapping = match &self.enable_span_mapping {
+            true => Some(true),
+            false => None,
+        };
```
Comment (bnjjj) [inline apollo-router/src/plugins/telemetry/tracing/datadog.rs:67]:
```suggestion
        let enable_span_mapping = self.enable_span_mapping.then_some(true);
```

**Review by bnjjj — COMMENTED** [review]
Could you add some documentation [here ](https://github.com/apollographql/router/blob/dev/docs/source/configuration/tracing.mdx#using-datadog) please ? To explain what it does and what's the difference.

Also could you generate your changelog using `cargo xtask changeset create`  please ?

Thanks a lot it looks great !

File: .changesets/config_datadog_exporter_mapping.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
+### Custom OpenTelemetry Datadog exporter mapping ([Issue #2228](https://github.com/apollographql/router/issues/2228))
+
+This PR fixes the issue with DD exporter not providing meaningful data in the DD traces.
```
Comment (tapaderster) [inline .changesets/config_datadog_exporter_mapping.md:3]:
```suggestion
This PR fixes the issue with the DataDog exporter not providing meaningful contextual data in the DataDog traces.
```

File: .changesets/config_datadog_exporter_mapping.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+There is a [known issue](https://docs.rs/opentelemetry-datadog/latest/opentelemetry_datadog/#quirks) where open telemetry is not fully compatible with datadog.
+
+To fix, this, open-telemetry-datadog added [custom mapping functions](https://docs.rs/opentelemetry-datadog/0.6.0/opentelemetry_datadog/struct.DatadogPipelineBuilder.html#method.with_resource_mapping).
```
Comment (tapaderster) [inline .changesets/config_datadog_exporter_mapping.md:6]:
```suggestion
To fix this, opentelemetry-datadog added [custom mapping functions](https://docs.rs/opentelemetry-datadog/0.6.0/opentelemetry_datadog/struct.DatadogPipelineBuilder.html#method.with_resource_mapping).
```

File: .changesets/config_datadog_exporter_mapping.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        | apollo_router router                                                              |
+            | apollo_router supergraph                                                      |
+            | apollo_router query_planning  | apollo_router execution                       |
+                                                | apollo_router fetch                       |
+                                                    | apollo_router subgraph                |
+                                                        | apollo_router subgraph_request    |
+```
+
+As you can see, there is no clear information about the name of the Query, the name of the Subgraph, and name of Query
+sent to Subgraph, when `enable_span_mapping` the following trace will be created:
```
Comment (tapaderster) [inline .changesets/config_datadog_exporter_mapping.md:29]:
```suggestion
As you can see, there is no clear information about the name of the query, the name of the subgraph, and the name of the query sent to the subgraph.
Instead when `enable_span_mapping` is set to `true` the following trace will be created:
```

File: .changesets/config_datadog_exporter_mapping.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                                              | fetch fetch                                              |
+                                                  | subgraph my-subgraph-name                            |
+                                                      | subgraph_request MyQuery__my-subgraph-name__0    |
+```
+
+
+
+
+
+All this logic is gated behind a yaml configuration boolean `enable_span_mapping` which if enabled will take the values from the span attributes.
```
Comment (tapaderster) [inline .changesets/config_datadog_exporter_mapping.md:45]:
```suggestion
All this logic is gated behind the configuration `enable_span_mapping` which if set to `true` will take the values from the span attributes.
```

File: docs/source/configuration/tracing.mdx [hunk 6]
```diff
@@ -153,10 +153,57 @@ The Apollo Router can be configured to connect to either the default agent addre
...
+        | apollo_router router                                                              |
+            | apollo_router supergraph                                                      |
+            | apollo_router query_planning  | apollo_router execution                       |
+                                                | apollo_router fetch                       |
+                                                    | apollo_router subgraph                |
+                                                        | apollo_router subgraph_request    |
+```
+
+As you can see, there is no clear information about the name of the Query, the name of the Subgraph, and name of Query
+sent to Subgraph, when `enable_span_mapping` the following trace will be created:
```
Comment (tapaderster) [inline docs/source/configuration/tracing.mdx:194]:
```suggestion
As you can see, there is no clear information about the name of the query, the name of the subgraph, and the name of the query sent to the subgraph.
Instead when `enable_span_mapping` is set to `true` the following trace will be created:
```

**Review by bnjjj — APPROVED** [review]

**Comment by Meemaw** [comment]
This is great! Will this work with otlp exporter as well?

**Comment by bnjjj** [comment]
@Meemaw I don't think so, are you experiencing the same issue using the otlp exporter ? I think we can't configure it at the router level but maybe it should be available at the otlp datadog collector ?

**Comment by Meemaw** [comment]
Yes, seeing same spans at reported in the issue (#1 screenshoot) using otlp exporter. Wondering if that can be handled in the router as well.

**Comment by bnjjj** [comment]
This ability is not provided by the tracing lib we're using because it's otlp compliant and doesn't really know if it's plugged on DD or not. Did you try to use this https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml#L3592 ?

**Pull request merged**